Add a small PHP 8.0 fix

xvilo opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

I'm updating logic so that it does not get used on PHP 8.0+. In PHP 8.0 and later, PHP uses libxml versions from 2.9.0, which disabled XXE by default. libxml_disable_entity_loader() is now deprecated.

The current code will throw deprecation warnings/errors and break the code. As for as far as I understand it, it the same logic as it is now disabled by default so changing this value is not needed. And also reverting the change back to its original value. 

I have opted to wrap the function calls in an if statement that checks if the current PHP version is lower than PHP 8.0. This was the easiest way I could think of to update this code without touching too much of its current behavior.

I'm not entirely sure if this is (already) wanted, this has been tested on the latest PHP 8.0 release candidate as of today. 

Also, please do let me know if you have feedback!

EDIT: 

Please note that I did not check any other part of the code, I just noticed this when running a project on PHP 8 as a test and came across this code hit on some its pages.

xvilo committed:
Update Security.php

Add a PHP version condition check.

In PHP 8.0 and later, PHP uses libxml versions from 2.9.0, which disabled XXE by default. libxml_disable_entity_loader() is now deprecated. This will throw errors for essentially the same feature.

Signed-off-by: Sem Schilder <sem@tropical.email>

boesing commented:
Thanks @xvilo!

Could you put a little bit more time into this and provide all necessary changes to achieve PHP 8.0 support?
I've tried to add an easy checklist in #2 for this.

Our community would really benefit from your changes.

xvilo committed:
Update minimum php requirement and phpunit

Signed-off-by: Sem Schilder <sem@tropical.email>

xvilo committed:
Update travis.yml to use php 8

Signed-off-by: Sem Schilder <sem@tropical.email>

xvilo commented:
@boesing thanks for your quick response, if have done the following as of now:
- Added `^8.0` as a supported/required PHP version
- Dropped support for PHP `5.6` and `7.2` and lower. 
- Implemented the latest version of PHPUnit, and ran tests. This succeeded with the same output as before.
- Update `travis.yml` by adding php 8 to the matrix and removing all 'unsupported' versions. 

The last step in your mentioned issue is something I did not completely do, as I don't have time for this. At least one issue have been fixed initially.

xvilo committed:
Add OS option

Signed-off-by: Sem Schilder <sem@tropical.email>

xvilo commented:
It seems Travis is not triggered anymore, any idea why?

boesing commented:
@xvilo Yeah, laminas exceeded limits... We have to wait till the next "free period" starts. We are aware of this but actually didn't come up with a solution.

boesing commented on .travis.yml:
```diff
@@ -44,6 +47,9 @@ matrix:
     - php: 7.3
       env:
         - DEPS=latest
+    - php: 8.0
```
Please double check #2' checklist and compare with the current matrix.
For any version, we have `DEPS=lowest` and `DEPS=latest`.

Since `8.0` is not yet available on travis, we have to use `nightly` instead.
As we are dropping PHP <7.3 in `composer.json`, please also drop old versions from the matrix and ensure that *all* supported versions are being tested by travis.

This includes:
- PHP 7.3
- PHP 7.4
- PHP 8.0 (`nightly` in travis for now)

xvilo replied:
Yes, I did this initially but could test it due to travis not working. I thought at first this might have been an issue on my side. Will update this later on!

boesing commented on composer.json:
```diff
@@ -25,12 +25,13 @@
     },
     "require": {
-        "php": "^5.6 || ^7.0",
+        "php": "^7.3 || ^8.0",
```
Please use the `~8.0.0` constraint as stated in #2

boesing commented on composer.json:
```diff
@@ -25,12 +25,13 @@
     "require-dev": {
         "laminas/laminas-coding-standard": "~1.0.0",
-        "phpunit/phpunit": "^5.7.27 || ^6.5.8 || ^7.1.4"
+        "phpunit/phpunit": "^9.4",
```
By bumping phpunit, you definitely have to change almost all tests contain `setUp` and `tearDown` aswell as have to check for some assertion methods.

Did you execute tests locally? You might want to fix those tests even without travis. 👍

xvilo replied:
I did run the tests on PHP 8 and via `./vendor/bin/phpunit` so not sure what happened there if they would be incompatible. I will double-check and see if I did something wrong 👍

weierophinney replied:
@boesing There are two test cases in the suite, and neither use `setUp` or `tearDown`, so tests actually _do_ run!

@xvilo I'm doing updates and testing manually, and will push to your branch for final review shortly. Thanks for getting this started!

weierophinney committed:
qa: update CI test matrix

- Updates to only include 7.3, 7.4, and nightly.
- Include lowest and latest targets for each PHP version

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: use tilde version range for PHP constraint

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update tests to make use of new PHP features

- Indented here/nowdocs
- Scalar typehints and return typehints
- `::class` instead of strings

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: move CS checks to PHP 7.3

The version we're using currently has deprecation notices on PHP 7.4, and failures to execute on 8.0.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
I've run tests against each of 7.3, 7.4, and 8.0 locally, using both latest and lowest dependencies. Tests all pass, as do CS-checks.

weierophinney commented on test/MultibyteTest.php:
```diff
@@ -27,76 +32,79 @@ public function multibyteEncodings()
-    <methodName>retrieved: &pocdata;</methodName>
-</methodCall>
-XML;
+            <?xml version="1.0" encoding="{ENCODING}"?>
+            <!DOCTYPE methodCall [
+                <!ENTITY pocdata SYSTEM "file:///etc/passwd">
+            ]>
+            <methodCall>
+                <methodName>retrieved: &pocdata;</methodName>
+            </methodCall>
+            XML;
```
Just a note, in case anybody wonders why I did this: PHP 7.3 adds the ability to indent heredocs and nowdocs. In each case, any leading whitespace before the closing marker is considered indentation, and stripped from each line of the string. This makes them far more readable when declared in indented code, which was the whole point of the change.

froschdesign commented on phpunit.xml.dist:
```diff
@@ -1,17 +1,18 @@
-         colors="true">
+<phpunit 
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+    xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" 
```
We can use the local XML schema definition, otherwise we have to update it with each version..

xvilo commented:
Thanks for finishing it up @weierophinney
